src/Microsoft.Health.Fhir.Store.Utils/BatchExtensions.cs

```diff
@@ -0,0 +1,138 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Health.Fhir.Store.Utils
+{
+    public static class BatchExtensions
+    {
+        private static Task StartTask(Action action, bool longRunning = true)
+        {
+            var task = new Task(action, longRunning ? TaskCreationOptions.LongRunning : TaskCreationOptions.None);
+            task.Start(TaskScheduler.Default);
+            return task;
+        }
+
+        internal static void ParallelForEach<T>(IEnumerable<T> objects, int threads, Action<int, T> action, CancelRequest cancel = null)
+        {
+            ExecuteInParallelBatches(objects, threads, 1, (thread, batch) => { action(thread, batch.Item2.First()); }, null, cancel);
+        }
+
+        public static void ExecuteInParallelBatches<T>(IEnumerable<T> objects, int threads, int batchSize, Action<int, Tuple<int, IList<T>>> action, int? queueCapacity = null, CancelRequest cancel = null)
+        {
+            if (cancel == null)
+            {
+                cancel = new CancelRequest();
+            }
+
+            using (var queue = queueCapacity == null ? new BlockingCollection<Tuple<int, IList<T>>>() : new BlockingCollection<Tuple<int, IList<T>>>(queueCapacity.Value))
+            {
+                Action<int> workerAction = (thread) =>
+                {
+                    while (!cancel.IsSet && !queue.IsCompleted)
+                    {
+                        try
+                        {
+                            if (queue.TryTake(out var batch, 1000))
+                            {
+                                if (!cancel.IsSet)
+                                {
+                                    action(thread, batch);
+                                }
+                            }
+                        }
+                        catch (Exception)
+                        {
+                            cancel.Set();
+                            queue.CompleteAdding(); // This is to let main thread to throw if it is blocked on write by limited queue capacity
+                            throw;
+                        }
+                    }
+
+                    if (cancel.IsSet)
+                    {
+                        queue.CompleteAdding(); // This is to let main thread to throw if it is blocked on write by limited queue capacity in case of external cancel
+                    }
+                };
+
+                var workers = new List<Task> { StartTask(() => workerAction(0)) };
+
+                try
+                {
+                    var batchId = 0;
+                    var batchList = new List<T>();
+                    foreach (var obj in objects)
+                    {
+                        if (cancel.IsSet)
+                        {
+                            break;
+                        }
+
+                        batchList.Add(obj);
+                        if (batchList.Count == batchSize)
```

Comment:
I'd consider handling batch size = 1, to avoid looping.


                    if (batchSize==1)
                    {
                        batchList = objects.ToList();
                        AddToQueueWithTrapping(queue, batchId, batchList, cancel, workers, threads, workerAction);
                    }
                    else ...